src/algebra/lie/nilpotent.lean

```diff
@@ -323,6 +329,73 @@ end
 
 end lie_module
 
+namespace lie_submodule
+
+variables {N₁ N₂ : lie_submodule R L M}
+
+/-- The upper (aka ascending) central series.
+
+See also `lie_submodule.lcs`. -/
+def ucs (k : ℕ) : lie_submodule R L M → lie_submodule R L M :=
+centralizer^[k]
+
+@[simp] lemma ucs_zero : N.ucs 0 = N := rfl
+
+@[simp] lemma ucs_succ (k : ℕ) :
+  N.ucs (k + 1) = (N.ucs k).centralizer :=
+function.iterate_succ_apply' centralizer k N
+
+@[mono] lemma ucs_mono (k : ℕ) (h : N₁ ≤ N₂) :
+  N₁.ucs k ≤ N₂.ucs k :=
+begin
+  induction k with k ih, { simpa, },
+  simp only [ucs_succ],
+  mono,
+end
+
+lemma ucs_eq_self_of_centralizer_eq_self (h : N₁.centralizer = N₁) (k : ℕ) :
+  N₁.ucs k = N₁ :=
+by { induction k with k ih, { simp, }, { rwa [ucs_succ, ih], }, }
+
+/-- If a Lie module `M` contains a self-centralizing Lie submodule `N`, then all terms of the upper
+central series of `M` are contained in `N`.
+
+An important instance of this situation arises from a Cartan subalgebra `H ⊆ L` with the roles of
+`L`, `M`, `N` played by `H`, `L`, `H`, respectively. -/
+lemma ucs_le_of_centralizer_eq_self (h : N₁.centralizer = N₁) (k : ℕ) :
+  (⊥ : lie_submodule R L M).ucs k ≤ N₁ :=
+by { rw ← ucs_eq_self_of_centralizer_eq_self h k, mono, simp, }
+
+lemma lcs_le_ucs_iff (l k : ℕ) :
+  N₁.lcs l ≤ N₂.ucs k ↔ N₁.lcs (l + k) ≤ N₂ :=
+begin
+  revert l,
+  induction k with k ih, { simp, },
+  intros l,
+  rw [(by abel : l + (k + 1) = l + 1 + k), ← ih, ucs_succ, lcs_succ, top_lie_le_iff_le_centralizer],
+end
+
+lemma lcs_le_ucs_iff' (k : ℕ) :
+  N₁ ≤ N₂.ucs k ↔ N₁.lcs k ≤ N₂ :=
+by { convert lcs_le_ucs_iff 0 k, rw zero_add, }
```

Comment:
Maybe worth adding a `galois_connection` lemma here?

Reply:
Done.

Although I'm not sure if we ever use these `galois` connection lemmas, I guess they do provide value by encouraging some consistency of what goes on the LHS and what goes on the RHS. (In this case, I flipped a few lemmas to avoid a `.symm` in the proof of `gc_lcs_ucs`.)